### PR TITLE
elegant crash complete protection

### PR DESCRIPTION
**Pull request by lonelybud**
add missing changes from -

https://github.com/YimMenu/YimMenu/issues/3002#issuecomment-2093241238

https://github.com/YimMenu/YimMenu/commit/f44edb8#commitcomment-141873742

also someone tried to add these changes back here #3096

credits @themonsterdev

### CODE REVIEW

**Comment by lonelybud** [comment]
@themonsterdev it should be you making this pr btw

**Comment by lonelybud** [comment]
i also want to add break statements in the for loops below in this pr itself -
https://github.com/YimMenu/YimMenu/blob/ef70781926c25066ac0455883fc5d58ce2029df1/src/hooks/protections/can_apply_data.cpp#L210-L229

**Comment by themonsterdev** [comment]
> @themonsterdev it should be you making this pr btw

I have a completely modified yim, that's why I don't do pr sorry but I have given you the protections

**Comment by gir489returns** [comment]
> > @themonsterdev it should be you making this pr btw
> 
> I have a completely modified yim, that's why I don't do pr sorry but I have given you the protections

That's why Git has branches.

File: src/hooks/protections/script_event_handler.cpp [hunk 1]
```diff
@@ -215,12 +215,28 @@ namespace big
...
+
+			if (g.protections.script_events.rotate_cam && static_cast<eRemoteEvent>(args[3]) == eRemoteEvent::TSECommandRotateCam)
 			{
 				g.reactions.rotate_cam.process(plyr);
 				return true;
 			}
+
+			if (g.protections.script_events.sound_spam && static_cast<eRemoteEvent>(args[3]) == eRemoteEvent::TSECommandSound)
+			{
+				if (!plyr || plyr->m_invites_rate_limit.process())
```
Comment (Yimura) [inline src/hooks/protections/script_event_handler.cpp]: `m_invites_rate_limit`, this does not seem like the right rate limit object.

Comment (lonelybud) [reply src/hooks/protections/script_event_handler.cpp]: ok i will add another rate limiter

Comment (lonelybud) [reply src/hooks/protections/script_event_handler.cpp]: i dont have idea about ideal rate_limiter constructor arguments but will add just fine for protection

**Comment by lonelybud** [comment]
> I have a completely modified yim, that's why I don't do pr sorry but I have given you the protections

you deserve the credit for your work and you know your work better than anyone and it is good practice you know

File: src/services/players/player.hpp [hunk 2]
```diff
@@ -67,6 +67,7 @@ namespace big
 
 		rate_limiter m_host_migration_rate_limit{2s, 15};
 		rate_limiter m_play_sound_rate_limit{1s, 10};
+		rate_limiter m_play_sound_rate_limit_tse{1s, 5}; // todo: confirm the correct values
```
Comment (Yimura) [inline src/services/players/player.hpp]: Now all that remains is to test it properly.

**Review by Yimura — DISMISSED** [review]

**Comment by themonsterdev** [comment]
> > > @themonsterdev ce devrait être toi qui fais ce pr btw
> > 
> > 
> > J'ai un yim complètement modifié, c'est pourquoi je ne fais pas de pr désolé mais je vous ai donné les protections
> 
> C'est pour ça que Git a des branches.

I believe that in private you could not, my project is private

File: src/services/players/player.hpp (hunk 2 above)
Comment (Rxann) [reply src/services/players/player.hpp]: @themonsterdev any suggestions on proper rate limiting amounts and time?

**Comment by gir489returns** [comment]
> > > > @themonsterdev ce devrait être toi qui fais ce pr btw
> > > 
> > > 
> > > J'ai un yim complètement modifié, c'est pourquoi je ne fais pas de pr désolé mais je vous ai donné les protections
> > 
> > 
> > C'est pour ça que Git a des branches.
> 
> I believe that in private you could not, my project is private

That has nothing to do with how Git works.

**Comment by Rxann** [comment]
> > > > > @themonsterdev ce devrait être toi qui fais ce pr btw
> 
> > > > 
> 
> > > > 
> 
> > > > J'ai un yim complètement modifié, c'est pourquoi je ne fais pas de pr désolé mais je vous ai donné les protections
> 
> > > 
> 
> > > 
> 
> > > C'est pour ça que Git a des branches.
> 
> > 
> 
> > I believe that in private you could not, my project is private
> 
> 
> 
> That has nothing to do with how Git works.

I agree, branches are a fundamental feature of git and it does not rely on public and private repository status.

**Comment by lonelybud** [comment]
@themonsterdev do we have any fixed value for `args[7]`. I added two values bcz in both of your comments they differ

**Comment by themonsterdev** [comment]
> @themonsterdev do we have any fixed value for `args[7]`. I added two values bcz in both of your comments they differ

if (args[3] == -4640169 && args[7] == -36565476 && args[8] == -53105203)

**Comment by lonelybud** [comment]
> if (args[3] == -4640169 && args[7] == -36565476 && args[8] == -53105203)

so value in this one is incorrect https://github.com/YimMenu/YimMenu/commit/f44edb8#commitcomment-141873742

**Comment by themonsterdev** [comment]
> @themonsterdev do we have any fixed value for `args[7]`. I added two values bcz in both of your comments they differ

if (args[3] == -4640169 && args[7] == -36565476 && args[8] == -53105203)

> > si (args[3] == -4640169 && args[7] == -36565476 && args[8] == -53105203)
> 
> so value in this one is incorrect [f44edb8#engagementcomment-141873742](https://github.com/YimMenu/YimMenu/commit/f44edb8#commitcomment-141873742)

I have to test as soon as I can with a stand menu, I modified the values so that my crash is not identical to that of the stand I may therefore have modified the values by carelessness

**Comment by Rxann** [comment]
> > if (args[3] == -4640169 && args[7] == -36565476 && args[8] == -53105203)
> 
> 
> 
> so value in this one is incorrect https://github.com/YimMenu/YimMenu/commit/f44edb8#commitcomment-141873742

I doubt it because I got elegant crashed and it was blocked.

**Comment by themonsterdev** [comment]
@Rxann 
> I doubt it because I got elegant crashed and it was blocked.

have you also blocked audio spam?

**Comment by Rxann** [comment]
> @Rxann 
> 
> > I doubt it because I got elegant crashed and it was blocked.
> 
> 
> 
> have you also blocked audio spam?

Not in YimMenu currently, that is part of what is being added in this PR

**Comment by themonsterdev** [comment]
> > @Rxann
> > > I doubt it because I got elegant crashed and it was blocked.
> > 
> > 
> > have you also blocked audio spam?
> 
> Not in YimMenu currently, that is part of what is being added in this PR

ok well that's why you crash, you also have to limit the spam of the sound that I gave ;)

**Comment by Rxann** [comment]
> > > @Rxann
> 
> > > > I doubt it because I got elegant crashed and it was blocked.
> 
> > > 
> 
> > > 
> 
> > > have you also blocked audio spam?
> 
> > 
> 
> > Not in YimMenu currently, that is part of what is being added in this PR
> 
> 
> 
> ok well that's why you crash, you also have to limit the spam of the sound that I gave ;)

What are the limits of the sound spam that you found useful?

**Comment by themonsterdev** [comment]
@Rxann 

```c++
if (g.protections.scripted_game_events.sound_spam && !NETWORK::NETWORK_IS_ACTIVITY_SESSION())
{
	if (static_cast<eRemoteEvent>(args[3]) == eRemoteEvent::TSECommandSoundSpam)
	{
		if (!plyr || plyr->m_invites_rate_limit.process())
		{
			if (plyr->m_invites_rate_limit.exceeded_last_process())
				g.reactions.sound_spam.process(plyr);

			return true;
		}
	}
}
```

```c++
rate_limiter m_invites_rate_limit{10s, 2};
```

**Comment by maybegreat48** [comment]
Are you sure that this protection is necessary? I've tested all Stand crashes a few days ago and none of them crash YimMenu

**Comment by tyackman** [comment]
> Are you sure that this protection is necessary? I've tested all Stand crashes a few days ago and none of them crash YimMenu

I was just crashed by a stand user with elegant(while using these changes). No stack trace produced.

EDIT: Maybe a fluke? I tried again and game survived. Freaky.

EDIT 2: Game shit the bed this time.

**Comment by rkwapisz** [comment]
> Are you sure that this protection is necessary? I've tested all Stand crashes a few days ago and none of them crash YimMenu

This is really the comment that was needed. Every one of my crashes the past few weeks I've been able to trace down into the menu myself (and 99.9% of them were my own mistakes). I fly around in an Op Mk 2 firing missiles at people way too often and see YimMenu blocking all manner of crash attempts without issue.

If there was a Stand crash that hits YimMenu, we'd be seeing it much more often. We're chasing a ghost here.

**Comment by abriginets** [comment]
> Are you sure that this protection is necessary? I've tested all Stand crashes a few days ago and none of them crash YimMenu

Apparently not https://github.com/YimMenu/YimMenu/issues/3106. Pretty much the same as @tyackman, one Stand user couldn't crash me, the other one succeeded. Either Stand was updated, or current protections are flaky.

**Comment by Rxann** [comment]
I think the amount of crashes stand has is underestimated. We could be experiencing similar but different crashes and not know it.

**Comment by gir489returns** [comment]
> I think the amount of crashes stand has is underestimated. We could be experiencing similar but different crashes and not know it.

It might also use progenartive crashes. @maybegreat48 has noted he isn't able to be crashed, that could be his account is marked by the Stand userbase not to use the "best" crashes it has against him. It could see that the user is still in the lobby after attempting the crash 10 times, then progress to a more difficult to handle crash, or a crash it might think we don't have blocked. However, I strongly doubt Stand is that good, and a lot of "ghost stories" of people who just claim to have been crashed by Stand have 0 idea who crashed them and with what menu, they just assume "LE HEBBIN STAND XD" and file a ticket without knowing any of the details, and then just expect us to handle it, like we know everything.

**Comment by abriginets** [comment]
In my own defense, if you are talking about me, I do not expect an instant silver bullet fix. If I knew C++ I would definitely go for a debugging session and create a PR instead of blasting issues at you guys. All I can do now is being proactive by reporting issues I'm having to try and make this menu better in terms of user experience. Peace ✌️

**Comment by lonelybud** [comment]
well i asked someone using stand with latest update to crash me and heres the reaction from my fork-

```
[19:56:11.8884100][WARN/reaction.cpp](111:5) Received Crash: Player ped removal from 'XXXX'
[19:56:11.9683882][WARN/reaction.cpp](111:5) Received Crash: invalid weapon type from 'XXXX'
[19:56:12.1781148][WARN/reaction.cpp](111:5) Received Crash: invalid weapon type from 'XXXX'
```

and it played audio as well. Despite adding sound spam changes in this pr i heard audio. Also there was no elegant crash notification which was #3076 . So these sound spam changes are not working fine or we need to set correct rate limiter values. I  did nt crashed. He tried 3 times.

In short the sound spam does not seem to crash and was not blocked even with these changes
Edit: i was host and he was my friend

**Comment by Yimura** [comment]
@lonelybud, how often did you hear the audio? Because the rate limiter "rate limits", it doesn't outright block only when a modder sends more than 2 events within 5 seconds (for example for this one).

**Comment by lonelybud** [comment]
> @lonelybud, how often did you hear the audio? Because the rate limiter "rate limits", it doesn't outright block only when a modder sends more than 2 events within 5 seconds (for example for this one).

the sound happened for less than a second. I should have logged all the events missed the opportunity :(

**Comment by themonsterdev** [comment]
I can confirm that the modifications that I gave work very well, the stand users do nothing to me at all, the modifications will let one sound or two pass but not more if it spams so it works well on my side.

In addition, you are assured of blocked crash 3 events regardless of the args

**Comment by themonsterdev** [comment]
incorrect !

```c++
case eRemoteEvent::Crash3:
		{
			if (isnan(*(float*)&args[4]) || isnan(*(float*)&args[5]))
			{
				g.reactions.crash.process(plyr);
				return true;
			}
			if (args[3] == -4640169 && args[7] == -36565476 && args[8] == -53105203)
			{
				session::add_infraction(plyr, Infraction::TRIED_CRASH_PLAYER);
				g.reactions.crash.process(plyr);

				return true;
			}
			break;
		}
```

correct !

```c++
case eRemoteEvent::Crash3:
{
	g.reactions.crash.m_event_name = "Crash 3.1";

	if (isnan(*(float*)&args[4]) || isnan(*(float*)&args[5]))
	{
		g.reactions.crash.m_event_name = "Crash 3.2";
		g.reactions.crash.process(plyr);
	}
	
	if (args[3] == -4640169 && args[7] == -36565476 && args[8] == -53105203)
	{
		g.reactions.crash.m_event_name = "Crash 3.3 (elegant)";
		g.reactions.crash.process(plyr);

		session::add_infraction(plyr, Infraction::TRIED_CRASH_PLAYER);
	}

	g.reactions.crash.m_event_name = "Crash";
	return true;
	break;
}
```

Also, be aware that sound spamming on PCs that are powerful won't work, which is why some don't crash and others do. block spam and crash 3 solved the problem for everyone!

**Comment by gir489returns** [comment]
> correct !
> 
> ```c++
> 	return true;
> 	break;
> ```

tenor

**Comment by themonsterdev** [comment]
it can happen to forget

**Comment by gir489returns** [comment]
> it can happen to forget

We're not even going to touch the issue of what I was making fun of, but what's the purpose of even checking the arguments, if your "happy path" is to just ignore the event anyway?

**Comment by themonsterdev** [comment]
> > it can happen to forget
> 
> We're not even going to touch the issue of what I was making fun of, but what's the purpose of even checking the arguments, if your "happy path" is to just ignore the event anyway?

well, to answer you, it serves two purposes for me;
1) make critics speak like you
2) maybe to detect the stands!

but hey if it's to take criticism like that I don't help more

**Comment by rkwapisz** [comment]
> incorrect !
> 
> ```c++
> case eRemoteEvent::Crash3:
> 		{
> 			if (isnan(*(float*)&args[4]) || isnan(*(float*)&args[5]))
> 			{
> 				g.reactions.crash.process(plyr);
> 				return true;
> 			}
> 			if (args[3] == -4640169 && args[7] == -36565476 && args[8] == -53105203)
> 			{
> 				session::add_infraction(plyr, Infraction::TRIED_CRASH_PLAYER);
> 				g.reactions.crash.process(plyr);
> 
> 				return true;
> 			}
> 			break;
> 		}
> ```
> 
> correct !
> 
> ```c++
> case eRemoteEvent::Crash3:
> {
> 	g.reactions.crash.m_event_name = "Crash 3.1";
> 
> 	if (isnan(*(float*)&args[4]) || isnan(*(float*)&args[5]))
> 	{
> 		g.reactions.crash.m_event_name = "Crash 3.2";
> 		g.reactions.crash.process(plyr);
> 	}
> 	
> 	if (args[3] == -4640169 && args[7] == -36565476 && args[8] == -53105203)
> 	{
> 		g.reactions.crash.m_event_name = "Crash 3.3 (elegant)";
> 		g.reactions.crash.process(plyr);
> 
> 		session::add_infraction(plyr, Infraction::TRIED_CRASH_PLAYER);
> 	}
> 
> 	g.reactions.crash.m_event_name = "Crash";
> 	return true;
> 	break;
> }
> ```
> 
> Also, be aware that sound spamming on PCs that are powerful won't work, which is why some don't crash and others do. block spam and crash 3 solved the problem for everyone!

Why are you bothering to do anything with g.reactions.crash.m_event_name if you're just going to assign it "Crash" 100% of the time?

**Comment by themonsterdev** [comment]
@rkwapisz 

g.reactions.crash.m_ event_name is used elsewhere if you looked, to not create or modify another reaction I use this one but do whatever you want as long as the crash is blocked

**Comment by gir489returns** [comment]
> well, to answer you, it serves two purposes for me;
> 
> 1. make critics speak like you
> 2. maybe to detect the stands!

> 1 I’m only pretending to be retarded!
> 2 I have no idea what I’m doing, and I get upset when called on it! It’ll stop the big bad boogeyman maybe with no evidence!!!!

Don’t help then.

**Comment by themonsterdev** [comment]
@gir489returns 

I'm not a dog, if you criticize me badly I react at the same level of stupidity, now if it's constructive criticism and questioned kindly why not

in your message this was not the case!

I prefer to avoid criticism, especially when it is mean and unconstructive, especially after spending several days looking for the solution. I've shared two approaches that have worked for me to help. Please respect my efforts rather than criticize them.

**Comment by gir489returns** [comment]
> in your message this was not the case!

You still haven't answered my question without being a little baby. Why even check the parameters at all? You're literally wasting the CPU's time.

**Comment by lonelybud** [comment]
> @lonelybud, how often did you hear the audio? Because the rate limiter "rate limits", it doesn't outright block only when a modder sends more than 2 events within 5 seconds (for example for this one).

heres the log where he tried elegant crash two times -
[log.zip](https://github.com/YimMenu/YimMenu/files/15322117/log.zip)

edit: i masked his name with `standuser1234`

**Comment by themonsterdev** [comment]
> > in your message this was not the case!
> 
> You still haven't answered my question without being a little baby. Why even check the parameters at all? You're literally wasting the CPU's time.

it's not a if that will change much

**Comment by gir489returns** [comment]
> > > in your message this was not the case!
> > 
> > 
> > You still haven't answered my question without being a little baby. Why even check the parameters at all? You're literally wasting the CPU's time.
> 
> it's not a if that will change much

That's not an argument.

**Comment by themonsterdev** [comment]
@gir489returns 
if you say so, moreover it allows me to detect the crash of the stand as I say

**Comment by lonelybud** [comment]
> > @lonelybud, how often did you hear the audio? Because the rate limiter "rate limits", it doesn't outright block only when a modder sends more than 2 events within 5 seconds (for example for this one).
> 
> heres the log where he tried elegant crash two times - [log.zip](https://github.com/YimMenu/YimMenu/files/15322117/log.zip)
> 
> edit: i masked his name with `standuser1234`

i think the reason why i did not blocked the sound bcz in my fork i have this rate limiter - `m_play_sound_rate_limit_tse{1s, 5}` and then in the logs the sound event happened 5 times under 1 second so it is obvious i will hear the sound

**Comment by themonsterdev** [comment]
@lonelybud 

m_play_sound_rate_limit_tse limits the sounds and does not block them so you will always hear sound according to the time you set

**Comment by lonelybud** [comment]
> @lonelybud
> 
> m_play_sound_rate_limit_tse limits the sounds and does not block them so you will always hear sound according to the time you set

yes i heard 5 times and if they were more than i should have logged sound spam in console but it never happened looks like stand only send 5 sound events?

**Comment by themonsterdev** [comment]
@lonelybud 

to my knowledge the stand sends the sounds 32 times, I test when I can again to see if there have been any changes

**Comment by lonelybud** [comment]
i think this pr is ready to be merged now

**Review by gir489returns — APPROVED** [review]

**Review by Yimura — APPROVED** [review]

**Pull request merged**